Source/FakeItEasy/Core/FakeScope.cs

```diff
@@ -13,19 +16,48 @@ namespace FakeItEasy.Core
     internal abstract class FakeScope
         : IFakeScope
     {
+        private const string CurrentScopeLogicalCallContextKey = "FakeItEasy.FakeScope.Current";
+
+        private static readonly RootScope root;
+        private static readonly ConcurrentDictionary<Guid, FakeScope> scopes;
+
+        [SuppressMessage(
+            "Microsoft.Performance", 
+            "CA1810:InitializeReferenceTypeStaticFieldsInline", 
+            Justification = "RootScope constructor requires scopes to be initialized first.")]
         static FakeScope()
         {
-            Current = new RootScope();
+            scopes = new ConcurrentDictionary<Guid, FakeScope>();
+            root = new RootScope();
         }
 
         private FakeScope()
         {
+            this.Id = Guid.NewGuid();
+            FakeScope.scopes.TryAdd(this.Id, this);

```

Comment:
nit: scopes.TryAdd ? (I am not at a compiler, so can't check). Similarly elsewhere.